### ISSUE TITLE
[5.2] Contextual binding does not work for Filesystem

### ISSUE THREAD

**Issue by hasyimibhar** [laravel/framework#12402]
It seems that I can't use contextual binding to override the default provided `Illuminate\Contracts\Filesystem\Filesystem` (`v5.2.10`):

```php
<?php namespace App;

use Illuminate\Contracts\Filesystem\Filesystem;

class TestFilesystem implements Filesystem
{
    // ...
}
```

```php
<?php

namespace App\Providers;

use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        //
    }

    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->when('App\Http\Controllers\TestController')
            ->needs('Illuminate\Contracts\Filesystem\Filesystem')
            ->give('App\TestFilesystem');
    }
}
```

```php
<?php

namespace App\Http\Controllers;

use Illuminate\Contracts\Filesystem\Filesystem;

class TestController extends Controller
{
	public function __construct(Filesystem $filesystem)
	{
		dd($filesystem);
	}
}
```

Output:

```php
FilesystemAdapter {#120 ▼
  #driver: Filesystem {#119 ▼
    #adapter: Local {#118 ▼
      #pathSeparator: "/"
      #permissionMap: array:2 [▼
        "file" => array:2 [▼
          "public" => 484
          "private" => 448
        ]
        "dir" => array:2 [▼
          "public" => 493
          "private" => 448
        ]
      ]
      #writeFlags: 2
      -linkHandling: 2
      #pathPrefix: "/Users/hasyimibahrudin/Projects/fstest/storage/app/"
    }
    #plugins: []
    #config: null
  }
}
```

Is this the expected behavior?

**Comment by GrahamCampbell** [comment]
What happens if you out that in boot out of interest?

**Comment by hasyimibhar** [comment]
The same thing happens. I still got `FilesystemAdapter` instead of `TestFilesystem`.

**Comment by cryode** [comment]
This is due to the `Illuminate\Contracts\Filesystem\Filesystem` interface being aliased to `filesystem.disk` by Laravel. By the time the contextual lookup is done, it is checking to see if `filesystem.disk` has a contextual setting, rather than the interface itself.

The quickest solution is to use `filesystem.disk` as the `needs()` parameter:

```php
/**
 * Register any application services.
 *
 * @return void
 */
public function register()
{
    $this->app->when('App\Http\Controllers\TestController')
        ->needs('filesystem.disk')
        ->give('App\TestFilesystem');
}
```

I wouldn't consider this a bug, just a conflict with the aliasing system (a definite frustration for the end user still, though). I don't know enough about the intention behind aliases to determine how this can be potentially resolved for a future version of Laravel. Personally, I don't like using anything other than fully qualified class names in the container.

**Comment by GrahamCampbell** [comment]
Thanks @cryode.

**Comment by taylorotwell** [comment]
I believe this should be fixed in upcoming Laravel 5.4 release.
